src/cli.rs

```diff
@@ -1,3 +1,558 @@
-pub fn run() {
-    println!("Hello, world!");
+#[cfg(test)]
+use mockall::automock;
+
+use crate::adapters::operations::device::DeviceOperations;
+use crate::core::operations::Operations;
+use crate::devices::local_file_storage::{
+    LocalFileStorage, StandardFileSystem, StandardPathProvider,
+};
+use crate::devices::mounted_folder::MountedFolderFactory;
+use crate::models::question::QuestionType;
+use crate::models::secondary_device::DeviceFactoryKey;
+use std::rc::Rc;
+
+const HELP: &str = r#"
+HibernaCLI
+Usage:
+    hibernacli [command] [options]
+
+Commands:
+    help              Display this help message
+    --version or -v   Display the version of the application
+    device [opt]      Manage devices
+    config [opt]      Manage configuration
+"#;
+
+const INVALID_COMMAND: &str = "Invalid command, use 'help' to display available commands";
+const VERSION: &str = env!("CARGO_PKG_VERSION");
+const DEFAULT_CONFIG: &str = "";
+const NEW_DEVICE_INIT: &str = "Create a new device";
+
+#[cfg_attr(test, automock)]
+trait UserInterface {
+    fn write(&self, message: &str) -> Result<(), String>;
+    fn read(&self) -> Result<String, String>;
+}
+
+struct Console;
+
+impl UserInterface for Console {
+    fn write(&self, message: &str) -> Result<(), String> {
+        println!("{}", message);
+        Ok(())
+    }
+    fn read(&self) -> Result<String, String> {
+        let mut buffer = String::new();
+        match std::io::stdin().read_line(&mut buffer) {
+            Ok(_) => return Ok(buffer),
+            Err(e) => return Err(e.to_string()),
+        };
+    }
+}
+
+struct CommandRunner<T: UserInterface, U: DeviceOperations> {
+    console: T,
+    operations: U,
+}
+
+impl<T: UserInterface, U: DeviceOperations> CommandRunner<T, U> {
+    fn run(&mut self, args: Vec<String>) {
+        if args.len() < 2 {
+            self.display_invalid_command();
+            return;
+        }
+
+        match args[1].as_str() {
+            "help" => self.display_help(),
+            "--version" | "-v" => self.display_version(),
+            "device" => self.run_device_command(args),
+            _ => {
+                self.display_invalid_command();
+            }
+        }
+    }
+
+    fn display_message(&self, message: &str) {
+        self.console.write(message);
+    }
+
+    fn read_string(&self) -> Result<String, String> {
+        self.console.read()
+    }
+
+    fn read_number(&self) -> Result<i32, String> {
+        let input = self.console.read()?;
+        input
+            .trim()
+            .parse::<i32>()
+            .map_err(|_| "Invalid number".to_string())
+    }
+
+    fn ask_question(&self, question_type: &QuestionType, question_statement: &str) -> String {
+        match question_type {
+            QuestionType::String => self.ask_for_string(question_statement),
```

Comment:
Warning: the only device we have for now also have a UnixPath question: 

```rs
impl MountedFolderFactory {
    pub fn new() -> MountedFolderFactory {
        MountedFolderFactory {
            path_question: Question::new(
                "What is the path of the folder?".to_string(),
                QuestionType::UnixPath,
            ),
            name_question: Question::new(
                "How would you name this device?".to_string(),
                QuestionType::String,
            ),
            step: 0,
        }
    }
```

Reply:
what level of verification should i do ? regex or actually check the path exists or nothing ?